### ISSUE TITLE
Content-Length header should be checked against max content length when uploading a file

### ISSUE THREAD

**Issue by mrubin** [playframework/playframework#8336]
I am on Play 2.6.x. Raising an issue as instructed at the Play forum - https://discuss.lightbend.com/t/behavior-when-uploading-a-multipart-file-which-exceeds-max-content-length-config/557

When I use nginx to enforce a max file size on file uploads and I upload a file that’s too large, nginx early-outs immediately (without uploading the entire file) - it must be looking at the `Content-Length` header.

I want to have Play enforce the max file size. However, when I disable the checking in nginx and allow Play to handle this limit, the limit does get enforced, but the entire file is uploaded to my server. This is not desirable.

My nginx proxy buffering is off.

I have confirmed that in the happy path, my controller method does receive the `Content-Length` header (it’s not being eaten by nginx or anything like that - it’s there). However, in the case that the file is too big, my controller method does not even get called and I see the following in the output:

```
2018-04-02 20:45:29,705 [ERROR] from akka.stream.impl.io.FileSubscriber in application-akka.actor.default-dispatcher-10 - Tearing down FileSink(/tmp/playtemp8312801549160061678/multipartBody6640999551180422187asTemporaryFile) due to upstream error
play.api.mvc.BodyParsers$MaxLengthLimitAttained: null
```

When I enable `log-config-on-start = true`, I see the following:

```
# This setting is set in `akka.http.server.parsing.max-content-length`
# Play uses the concept of a `BodyParser` to enforce this limit, so we override it to infinite.
"max-content-length" : "infinite",
```

So Akka is not checking the limit (maybe it would use the `Content-Length` and early out without allowing the entire file to be uploaded). I tried to set `play.http.server.max-content-length = 10m` in my application.conf, but it looks like Play overrides it. I was not able to change this value.

Looking at https://github.com/playframework/playframework/blob/b5a88b3af2b806d519cf4d7afad704f847f4899d/framework/src/play/src/main/scala/play/api/mvc/BodyParsers.scala and how `DefaultMaxDiskLength` is used,

```
val takeUpToFlow = Flow.fromGraph(new BodyParsers.TakeUpTo(maxLength))
```

I don’t see anything making use of `Content-Length` in any of these (body parser) files.

It seems desirable for checking `Content-Length`, if it’s there, to be the default behavior, as it is with nginx. Why allow an attacker to upload a gigabyte-sized file to your server only to have it be rejected for being over 10 megs? Is it possible to use the `Content-Length` on the request and early-out with a 413 without allowing the entire file to be uploaded?

**Comment by mkurz** [comment]
> I tried to set `play.http.server.max-content-length = 10m` in my application.conf

That settings does not exist. What you [are looking for](https://github.com/playframework/playframework/blob/2.7.0-RC3/framework/src/play-akka-http-server/src/main/resources/reference.conf#L55) is
```
play.server.akka.max-content-length = 10MB
```
**But be aware:** This setting will **not** be picked up by akka in dev mode (but in production when staging your app for example it will).

However you can also apply this setting in dev mode by adding this to your `build.sbt`:
```
PlayKeys.devSettings += "play.server.akka.max-content-length" -> "10MB"
```

However, I think body parsers should check the `Content-Length` as well.

**Comment by mkurz** [comment]
What I don't like about `play.server.akka.max-content-length = ...` is that when the size exceeds the server returns 500 instead of 413. Also it would be nice if the error handler would be called...

**Comment by marcospereira** [comment]
+1 for returning a `413` instead of a `500`.

**Comment by wyb008** [comment]
Hi everynone. Appreciate you can help on my Big File upload issue.
We need to allow the user to upload big files(hundred MB or even GB) to our PLAY Server. I have added the mentioned congifurations in my application.conf:
play.server.akka.max-content-length = 100MB
play.http.parser.maxMemoryBuffer = 100MB
parsers.MultipartFormData.maxLength = 100MB
parsers.formUrlEncoded.maxLength=102400K
akka.http.server.parsing.max-content-length = 102400K

And my Controller code:
def uploadFiles(path: String) = Action(parse.multipartFormData){ implicit request =>
    val dir = request.getQueryString("path").get
    val root = new File(rootPath + dir)
    val files = request.body.files

    files.map( file => {
      val filename    = Paths.get(file.filename).getFileName
      file.ref.copyTo(Paths.get(s"${root}\\$filename"), replace = true)
    })

    Ok("create successfully")
  }

Then, it always throw the exception:
[error] a.s.i.i.FileSubscriber - Tearing down FileSink(C:\Users\admin\AppData\Local\Temp\playtemp7508536138098091876\multipartBody8797152134196752356asTemporaryFile) due to upstream error
play.api.mvc.BodyParsers$MaxLengthLimitAttained: null

**Comment by mkurz** [comment]
> +1 for returning a `413` instead of a `500`.

Will be fixed by #9625

**Comment by mkurz** [comment]
@wyb008 @maljukan A bit late, but I guess all you need to configure is:
```
play.http.parser.maxDiskBuffer = 1GB
```
Nothing more, you can remove all your other settings (actually you _should_ remove `play.server.akka.max-content-length` or set it at least at the same value like `maxDiskBuffer`).
This is actually [documented](https://www.playframework.com/documentation/2.7.x/ScalaBodyParsers#Max-content-length): File uploads via `multipart/form-data` will use a parsers that buffer content on disk, and therefore the `maxDiskBuffer` setting is relevant here (not the `maxMemoryBuffer` one).

BTW: The last three config you listed
```
parsers.MultipartFormData.maxLength = 100MB
parsers.formUrlEncoded.maxLength=102400K
akka.http.server.parsing.max-content-length = 102400K
```
**do not even exist!**

**Comment by maljukan** [comment]
@mkurz I've tried to set only:
```
play.server.akka.max-content-length=1GB
play.http.parser.maxDiskBuffer = 1GB
```
But when i try to do an upload via `multipart/form-data` i always get:
`Error: Request Entity Too Large (413)`
Some default limit somewhat under 10MB is always applied no matter what i've tried following the documentation for **play 2.6.x** and also AkkaHttpServer. Tried all possible combinations like `g GB m MB k KB` but nothing worked out.

**Comment by mkurz** [comment]
@maljukan 
- First: How large is the file you want to upload? 
- Second: Are you running your app behind a reverse proxy (nginx,..)? Maybe this one already catches a big file upload and returns 413 before even passing the requeest to your Play app?
- Third: Be aware that **in dev mode** you have to set any `play.server.*` configs in `build.sbt` via `PlayKeys.devSettings` to take effect:
```
PlayKeys.devSettings += "play.server.akka.max-content-length" -> "1GB"
```
All `play.server.*` settings _configured in `application.conf`_ will only be picked up in prod mode.

**Comment by mkurz** [comment]
But actually you do not even have to set `play.server.akka.max-content-length` because it's `infinite` by default anyway...

**Comment by maljukan** [comment]
Thanks @mkurz 
* The file is about 23MB
* Yes, maybe it's caught by nginx although i've put `client_max_body_size 100MB` in `nginx.conf` and `sites-available/site.conf` I'll try to find a way to eliminate that possibility but... not sure how to validate this assumption atm
* It's a production and i've put `Dplay.server.akka.max-content-length=infinite` to avoid redeploying after removing that entry from `application.conf`

**Comment by maljukan** [comment]
@mkurz Oh i just saw you referenced the status issue fix. 
Great! Then it's something along the way. The server is a VM instance on the Google Cloud Platform. No load balancing but you pointed me to correct direction. Thanks for chewing it for me!

**Comment by mkurz** [comment]
@maljukan Let me know when you find the cause :wink:

**Comment by mkurz** [comment]
Partially fixed by #9625, will be completely fixed by #9649.
